### ISSUE TITLE
lxc list consumes ~15-20 % of cpu

### ISSUE THREAD

**Issue by upuvvad1** [lxc/lxd#2472]
lxd 2.4
lxc 2.4
ubuntu 16.04

LXC list consumes lots of CPU.
I am working on a project, which needs to poll lxd every 2 seconds and update list of containers.
I am doing exec of "lxc list" to get the list of containers and their state.
I thought of writing a query to poll from lxd.db, however, it contains only the names of the containers, but their state (running/stopped/frozen) wasn't  available in the DB.

Is there any file, which I can read to obtain the state of the container ?

 watch lxc list 

![image](https://cloud.githubusercontent.com/assets/17104752/19200439/014b280c-8c97-11e6-933f-6325eae00356.png)

**Comment by stgraber** [comment]
The reason why "lxc list" is so expensive is because it's a very very expensive operation.

On the client side to get half reasonable performance, the command hits at least 2 API endpoints for every listed container, it does that in parallel to make things as fast as possible.

On the server side, LXD needs to talk to LXC over a unix socket to query the state, look in the database for snapshots, inject code inside the container to pull its IP addresses out, ...

So yes, scripting around "lxc list" in a tight loop is probably not the best idea, or at least not with the default arguments for it.

What information is it that you want exactly?

**Comment by stgraber** [comment]
(And to answer the obvious question, the container can change state on its own without talking to LXD or LXD even running. The LXD database only stores information which LXD itself controls, the rest is retrieved from the source.)

**Comment by upuvvad1** [comment]
I am scripting to  monitor the host and container resources which are running, and want to trigger the migration if the container is consuming x% of host's resources to the one with high-end configuration.

I am reading the containers resources data from /sys/fs/cgroup/<resource>/lxc/<cnt-name> files

and using the "lxc list" to obtain the state of the containers.

As an alternative, I can populate data from db, and then see if container name found in /sys/fs/cgroup directory structure and make it work.
Was curious why it's consuming lots of resources.

In short, I am looking at the resource usage information, of all containers.

**Comment by stgraber** [comment]
Using - - fast should make a massive improvement for you

**Comment by upuvvad1** [comment]
Hmm.. 
Thanks for a clear explanation. :)
